primitives/src/crh/poseidon/updatable/mod.rs

```diff
@@ -0,0 +1,217 @@
+use super::*;
+use std::marker::PhantomData;
+use crate::UpdatableFieldBasedHash;
+
+pub struct UpdatablePoseidonHash<F: PrimeField + MulShort, P: PoseidonParameters<Fr = F>>{
+    state: Vec<F>,
+    pending: Vec<F>,
+    _parameters: PhantomData<P>,
+}
+
+impl<F, P> UpdatablePoseidonHash<F, P>
+    where
+        F: PrimeField + MulShort ,
+        P: PoseidonParameters<Fr = F>,
+{
+    pub fn new(personalization: Option<&[F]>) -> Self {
+        let mut state = Vec::with_capacity(P::T);
+        for i in 0..P::T {
+            state.push(P::AFTER_ZERO_PERM[i]);
+        }
+        let mut instance = Self {
+            state,
+            pending: Vec::with_capacity(P::R),
+            _parameters: PhantomData,
+        };
+
+        // If personalization Vec is not multiple of the rate, we pad it with zero field elements.
+        // This will allow eventually to precompute the constants of the initial state. This
+        // is exactly as doing H(personalization, padding, ...). NOTE: this way of personalizing
+        // the hash is not mentioned in https://eprint.iacr.org/2019/458.pdf
+        if personalization.is_some(){
+            let personalization = personalization.unwrap();
+            let padding = personalization.len() % P::R;
+
+            for p in personalization.iter(){
+                instance.update(p);
+            }
+
+            for _ in 0..padding {
+                instance.update(&F::zero());
+            }
+            assert_eq!(instance.pending.len(), 0);
```

Comment:
I think you mean: 
if padding != 0 then padding should be P::R - personalisation.len() % P::R